bump signal export to 3.2.2

belamu opened the pull request:
Update signal-export to version 3.2.0

## Why
signal-desktop has changed a lot since signal-export 1.8.2 and therefore signal-export changed as well.

## Things done

- update version to 3.2.0
- change pypi package name (`signal_export` instead of `signal-export` since otherwise the tarball is not found)
- add dependencies that where missing (see [pyproject.toml](https://github.com/carderne/signal-export/blob/main/pyproject.toml#L36))
- switch build-backend to pdm-backend (see [pyproject.toml](https://github.com/carderne/signal-export/blob/main/pyproject.toml#L58)

### Fixes missing
- dependency for sqlcipher is [sqlcipher3-wheels](https://github.com/carderne/signal-export/blob/main/pyproject.toml#L37) which is basically the same code as pysqlcipher3 but the nix build fails with 
```
Running phase: pythonRuntimeDepsCheckHook
@nix { "action": "setPhase", "phase": "pythonRuntimeDepsCheckHook" }
Executing pythonRuntimeDepsCheck
Checking runtime dependencies for signal_export-3.2.0-py3-none-any.whl
  - sqlcipher3-wheels not installed
```
I am not proficiant enough to package sqlcipher3**-wheels** and I am also unsure if it is a good idea since maybe the "wheels"-wrapping this provides should rather be done in nixpkgs at python312Packages.sqlcipher3 or in signal-export?
## Checks

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage) (not applicable, no usages)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [-] (Package updates) Added a release notes entry if the change is major or breaking (not major)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Commit:
signal-export: increase version to 3.2.0

phaer commented on pkgs/by-name/si/signal-export/package.nix:
```diff
@@ -6,16 +6,17 @@
+    pdm-backend
   ];
 
   propagatedBuildInputs = with python3.pkgs; [
```
we should be able to drop setuptools here, now that the package uses pdm-backend

belamu replied:
:+1: I remove it from the BuildInputs as well.

phaer commented on pkgs/by-name/si/signal-export/package.nix:
```diff
@@ -25,6 +26,8 @@ python3.pkgs.buildPythonApplication rec {
     emoji
     markdown
     pysqlcipher3
+    rich
```
Why do we need rich here? I don't see it in https://github.com/carderne/signal-export/blob/main/pyproject.toml

belamu replied:
:+1: 
We get it indirectly via typer. I found it in requirements.lock and didn't pay attention that it is an indirect dependency. I remove it.

phaer commented on pkgs/by-name/si/signal-export/package.nix:
```diff
@@ -25,6 +26,8 @@ python3.pkgs.buildPythonApplication rec {
     emoji
     markdown
     pysqlcipher3
```
I believe this is the wrong package: The unmaintained https://github.com/rigglemania/pysqlcipher3/.

The package wants "sqlcipher3-wheels", which lives at https://github.com/laggykiller/sqlcipher3 according to pypi and is itself a fork of https://github.com/coleifer/sqlcipher3 and seems to mainly add a workflow to build wheels and do some openssl-shenanigans. 

I suggest we try to package https://github.com/coleifer/sqlcipher3 directly and patch signal-export to use "sqlcipher3" instead of "sqlcipher3-wheels". I can do that but it might take me a few days.

phaer reviewed (commented):
Thank you so much for highlighting this update and delivering this draft PR! I started to watch upstream on github now, so I don't miss new releases in the future.

phaer committed:
python3Packages.sqlcipher3: init at 0.5.4

This adds python bindings for sqlcipher from
https://github.com/coleifer/sqlcipher3.

It also adds two meta packages, sqlcipher3-binary
and sqlcipher3-wheels. The former is a binary-distribution
of sqlcipher3 from upstream. The latter is a fork that
builds its own wheels but ships no further source changes.

So we effectively make both wheel packages an alias for our
own source-built package. Similar to how e.g. psycopg2-binary
works in nixpkgs.

phaer commented:
@belamu I have pushed two commits to my fork https://github.com/NixOS/nixpkgs/compare/master...phaer:nixpkgs:353082-bump-signal-export-to-3.2.0 . I packaged the right sqlcipher3 and signal-export seems to build. I can't currently test it easily, so please do if you can :)

belamu commented:
Wow, thanks. And quick reaction!
I'll try it out in a minute.
What I don't understand is: why did you create two MetaPackages with the exact same content (`sqlcipher3-binary` and `sqlcipher3-wheels`).

belamu commented:
Worked flawlessly! Now my backup routine is complete again after moving to nixOS :)

phaer commented:
> What I don't understand is: why did you create two MetaPackages with the exact same content (sqlcipher3-binary and sqlcipher3-wheels)

Because all 3 are effectively the same package, the later two are just distributed as binary wheels on pypi. As we are building the package from source anyway in nixpkgs, and one of the outputs is a wheel, we just reuse that. And update metadata to match the name, so things which depend on `sqlcipher3-wheels` instead of `sqlcipher3` build without further patching.

Does that make sense?

> Worked flawlessly! Now my backup routine is complete again after moving to nixOS :)

Great to hear! Then you could push those commits to your branch and undraft the PR?

belamu commented:
> Does that make sense?

It does to me.

> undraft the PR?

Yes. I combined the commits to decrease clutter in the history. The result is exactly the same as with your changes. Your dependency-fixing is not attributed to you (the new package is), I hope that's fine.

belamu commented:
> - [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. 

I couldn't run this because the git checkout or tarball extraction failed, maybe because of too little RAM. But there shouldn't be any packages this change depends on since `signal-export` is "end-user-software" and the sqlcipher is new.

picnoir commented:
Oh no. I screwed up ><

I was fixing the git commit message and did not see your push @belamu, sorry about that ><. Fixing this.

belamu commented:
@phaer in case you saw the new signal-export release: it's broken, I messed it up and will fix it in the next days
See the [pr](https://github.com/carderne/signal-export/pull/146#issuecomment-2458828503)

natsukium commented on pkgs/development/python-modules/sqlcipher3/default.nix:
```diff
@@ -0,0 +1,39 @@
+    hash = "sha256-4w/1jWTdQ+Gezt3RARahonrR2YiMxCRcdfK9qbA4Tnc=";
+  };
+
+  nativeBuildInputs = [
```
```suggestion
  build-system = [
```

belamu replied:
based on the terms this looks good.
Unfortunately I couldn't figure out by reading on different sites like the [the manual](https://nixos.org/manual/nixpkgs/unstable/#var-stdenv-buildInputs) what really the difference is.
I thing I've found "build-system" only in the pythonPackaging-context while nativeBuildInputs looks like something more general.
If anyone has time I would like to read why one is more preferable then the other (newbie speaking).

natsukium replied:
I'm sorry for not getting back to you sooner. I was overwhelmed by the notification.
You are correct that `build-system` is python builder-specific.

You can find more information in the python section, https://nixos.org/manual/nixpkgs/unstable/#handling-dependencies

belamu replied:
I've incorporated this suggested change in 1bae375e38b8f6a0a0da953d2b48ecf2bc629b20 . Therefore I hope, it's merge-ready(er).

natsukium commented on pkgs/development/python-modules/sqlcipher3/default.nix:
```diff
@@ -0,0 +1,39 @@
+    sqlcipher
+    openssl
+  ];
+
```
Could you enable tests with `pythonImportsCheck` and `pytestCheckHook`?

https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#testing-python-packages-testing-python-packages

belamu replied:
@phaer that was your code, would you mind to edit this? (Push to belamu:353082-bump-signal-export-to-3.2.0 should be possible for you)

belamu replied:
see below: pythonImportCheck done. pytestCheckHook not really feasible.

natsukium commented on pkgs/development/python-modules/sqlcipher3/default.nix:
```diff
@@ -0,0 +1,39 @@
+    mainProgram = "sqlcipher3";
+    homepage = "https://github.com/coleifer/sqlcipher3";
+    description = "Python 3 bindings for SQLCipher";
+    platforms = platforms.unix;
```
```suggestion
```
`buildPythonPackage` sets `meta.platforms` by default.

belamu replied:
@phaer that was your code, would you mind to edit this? (Push to belamu:353082-bump-signal-export-to-3.2.0 should be possible for you)

My question again (from a learning ambition): where could I've seen this information? The only documentation of buildPythonPackage I've found was [here](https://dream2nix.dev/reference/buildPythonPackage/) which did not write about meta(.platforms). So far I haven't l found/learned where buildPythonPackage (or similar) are defined to see this information in source code.

natsukium replied:
It is the dream2nix documentation, not the nixpkgs one.
The nixpkgs documentation is [here](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#using-pythonimportscheck-using-pythonimportscheck). (or [rendered one](https://nixos.org/manual/nixpkgs/unstable/#python))

Unfortunately, `meta` is not mentioned, but the implementation is here.
https://github.com/NixOS/nixpkgs/blob/f93401eeed99c853c8c92158555dcc17eae4d914/pkgs/development/interpreters/python/mk-python-derivation.nix#L326-L330

phaer replied:
> @phaer that was your code, would you mind to edit this? (Push to belamu:353082-bump-signal-export-to-3.2.0 should be possible for you)

Sorry, I've been busy and the invitation expired. Can you try adding the test hook and removing this line, before we go back and forth on this again? :)

belamu replied:
@phaer Sorry, I was not able to. Removing the meta line of course but the tests are more confusing for me.
What I figured out:
- tests for sqlcipher3 are supposed to be run [by calling `python tests/`](https://github.com/coleifer/sqlcipher3/blob/e2176211155d0cc0a6580a9d9788b84fd77ef254/.github/workflows/tests.yaml#L24C14-L24C27)
- the tests use `unittest` which is supposed to be [compatible](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#using-unittestcheckhook-using-unittestcheckhook) with `pytest`
- adding `nativeCheckInputs= [ pytestCheckHook ]; pytestFlagsArray = [ "tests/" ];` (and `pytestCheckHook` in the arguments at the top of `default.nix`) results in error because `tests` is not found
- the tar that is downloaded (in /nix/store/8232cdbi6vxb55l1hdvkk6f7xrf3p8np-sqlcipher3-0.5.4.tar.gz ) does not contain a directory `tests`
  - do we need to download a complete source?
- `pythonImportsCheck` seems to be needed for python packages that are imported. The imported modules are `datetime`, `time`, `collections.abc`. But to check them, they would need to be listed in the arguments at the top of `default.nix` I guess which seems unreasonable for standard modules. Not sure if the needed non-python-dependencies should be checked with `pythonImportsCheck`.

If you understand that better, I would be glad if you could fix this.

phaer replied:
Hello!

I just (force-)pushed to your branch.

Just added pythonImportChecks hook. That one checks that the exported module, sqliteciphers3 in this case, can be imported - which gives some indication that the build is not completely broken.

The problem with pytestCheckHook here is that the test files don't start with "test_". We would need to patch in a pytest.ini or so with "python_files" set, but then we run into a nixpkgs problem where the built python extension is only installed in the output, not in the build directory where the checkPhase runs by default.

We could fix that, but I'd prefer to do that in another PR if needed. You are also correct that we would have to download the source from github or so for that to work.

Commit:
signal-export: 3.2.0 -> 3.2.2

includes decryption with more key providers, namely the one from KDE

b12f commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 353086`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 14 packages built:</summary>
  <ul>
    <li>python311Packages.sqlcipher3</li>
    <li>python311Packages.sqlcipher3-binary</li>
    <li>python311Packages.sqlcipher3-binary.dist</li>
    <li>python311Packages.sqlcipher3-wheels</li>
    <li>python311Packages.sqlcipher3-wheels.dist</li>
    <li>python311Packages.sqlcipher3.dist</li>
    <li>python312Packages.sqlcipher3</li>
    <li>python312Packages.sqlcipher3-binary</li>
    <li>python312Packages.sqlcipher3-binary.dist</li>
    <li>python312Packages.sqlcipher3-wheels</li>
    <li>python312Packages.sqlcipher3-wheels.dist</li>
    <li>python312Packages.sqlcipher3.dist</li>
    <li>signal-export</li>
    <li>signal-export.dist</li>
  </ul>
</details>

belamu commented:
@natsukium 
From my point of view, the pull request is ready for merge even though unfortunately some tests are missing:
- for pysqlcipher3: as @phaer said it's complicated. Sorry, I only understand it rudimentarily
- for signal-export: the only test that the repo provides is outdated and broken, as I noted [here](https://github.com/carderne/signal-export/issues/150). Therefore it is not included. I tried it out with a current signal-desktop nixos installation on KDE and I hope that @carderne (Maintainer of signal-export) tried it out with Gnome.

I cannot build and try out on Darwin. There is platform dependent code that I did not touch during the last commits. It would be nice if someone tried but I hope this PR can be merged without it.

phaer commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 353086`

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 14 packages built:</summary>
  <ul>
    <li>python311Packages.sqlcipher3</li>
    <li>python311Packages.sqlcipher3-binary</li>
    <li>python311Packages.sqlcipher3-binary.dist</li>
    <li>python311Packages.sqlcipher3-wheels</li>
    <li>python311Packages.sqlcipher3-wheels.dist</li>
    <li>python311Packages.sqlcipher3.dist</li>
    <li>python312Packages.sqlcipher3</li>
    <li>python312Packages.sqlcipher3-binary</li>
    <li>python312Packages.sqlcipher3-binary.dist</li>
    <li>python312Packages.sqlcipher3-wheels</li>
    <li>python312Packages.sqlcipher3-wheels.dist</li>
    <li>python312Packages.sqlcipher3.dist</li>
    <li>signal-export</li>
    <li>signal-export.dist</li>
  </ul>
</details>

Commit:
signal-export: use built-system key

the built-system key is specific to python packages and therefore preferable over the more general nativeBuildInputs

see https://github.com/NixOS/nixpkgs/pull/353086#discussion_r1832644539

belamu commented:
> aarch64-darwin

Thanks for trying it out on MacOS as well.
Unfortunately I still cannot run it on Linux even, some git cache problem. Fortunately @b12f did it.

picnoir reviewed (approved):
LGTM

@belamu, I see this is one of your first contributions. First of all, congrats for jumping through this hoop!

This has been running 2 months for a bump (partly my fault), I wanted to say this is the somewhat the worst kind of situation you can end up, it usually go faster, I hope you won't leave with a bitter taste :)

belamu commented:
> @belamu, I see this is one of your first contributions. First of all, congrats for jumping through this hoop!
> 
> This has been running 2 months for a bump (partly my fault), I wanted to say this is the somewhat the worst kind of situation you can end up, it usually go faster, I hope you won't leave with a bitter taste :)

Thank you for your kind words! I did not have the impression of a bad situation but had a positive surprise: I did a small step and immediatly got a big contribution from you ( @phaer ). Since I didn't react to some of the reviews I attributed the delay not (or not much) to you.
Together with answers to [my question on Discourse](https://discourse.nixos.org/t/how-to-avoid-system-freeze-maybe-connected-to-swap-ram/56301/4) I had a very helpful introduction into the Nix community. Oh, not to forget the install support at a Nix meetup in Dresden and @b12f 's help at CCC in Hamburg.
